toolchain/lower/file_context.cpp

```diff
@@ -121,6 +137,219 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
   return std::move(llvm_module_);
 }
 
+auto FileContext::CoalesceEquivalentSpecifics() -> void {
+  for (auto& specifics : lowered_specifics_) {
+    // i cannot be unsigned due to the comparison with a negative number when
+    // the specifics vector is empty.
+    for (int i = 0; i < static_cast<int>(specifics.size()) - 1; ++i) {
+      if (is_replaced_specific_[specifics[i].index]) {
+        specifics[i] = specifics[specifics.size() - 1];
+        specifics.pop_back();
+        --i;
+        continue;
+      }
+      for (unsigned j = i + 1; j < specifics.size(); ++j) {
+        if (is_replaced_specific_[specifics[j].index]) {
+          specifics[j] = specifics[specifics.size() - 1];
+          specifics.pop_back();
+          --j;
+          continue;
+        }
+        std::pair<SemIR::SpecificId, SemIR::SpecificId> pair = {specifics[i],
+                                                                specifics[j]};
+        std::pair<SemIR::SpecificId, SemIR::SpecificId> flipped_pair = {
+            specifics[j], specifics[i]};
```

Comment:
Is it possible to create a canonical ordering here? e.g.:

```suggestion
        std::pair<SemIR::SpecificId, SemIR::SpecificId> pair = {specifics[i],
                                                                specifics[j]};
        if (pair.first.index > pair.second.index) {
          // Ensure IDs are sorted for queries.
          std::swap(pair.first.index, pair.second.index);
        }
```

Reply:
Or, what about sorting `specifics` by index before looping? That might actually give similar results. Especially if you can remove the swapping / pop_back replacements in `specifics`, you might be able to guarantee the ordering of the pair without needing to individually check (because of the order of the traversal).

Note, I bring this up mainly because I'd like to remove `visited_equivalent_specifics_flipped` -- I think it should be possible to remove it, and having two separate sets with similar data seems worth avoiding due to the hash overhead.